lib/wallet/walletdb.js

```diff
@@ -503,6 +503,105 @@ class WalletDB extends EventEmitter {
     }
   }
 
+  /**
+   * Deep rescan:
+   * Keep all keys, account data, wallet maps (name and path).
+   * Dump all TX history and balance state, rescan from genesis.
+   * @param {Boolean?} rescan -- only used for testing
+   * @returns {Promise}
+   */
+
+  async deepRescan(rescan = true) {
+    const unlock1 = await this.txLock.lock();
+    const unlock2 = await this.writeLock.lock();
+    const unlock3 = await this.readLock.lock();
+    try {
+      return await this._deepRescan(rescan);
+    } finally {
+      unlock3();
+      unlock2();
+      unlock1();
+    }
+  }
+
+  /**
+   * Deep rescan (without locks)
+   * Keep all keys, account data, wallet name and path maps.
+   * Dump all TX history and balance state, rescan from genesis.
+   * @param {Boolean?} rescan -- only used for testing
+   * @returns {Promise}
+   */
+
+  async _deepRescan(rescan) {
```

Comment:
Could break `_deepRescan` into two parts, drop and scan, then call those directly from the test, instead. 'Test only' params are just asking to be passed wrong.

Reply:
Yeah this is a good point, I'll think about how it could be done and if that's worth it.
You caught me in that I'm just using the param for testing but you could also think of it like the API just doesn't expose that option. I don't want the user to have to execute two different commands to fix their wallet state, and after the data is deleted a rescan is really required.

Reply:
Probably not worth the work, but the way I would do it is:

- Split `_deepRescan` by moving 590-602 to a new method. I'd call it something like `_rescanFromGenesis` or something.
- Remove the parameter from `_deepRescan` and rename it to '_dropHistory` or something
- Keep `deepRescan` as the top-level entry point, but remove the parameter and have it call `_dropHistory` then `_rescanFromGenesis`
- Tests just call the `_dropHistory` and `_rescanfromGenesis` at the appropriate time.

Might end up just making it harder to read.

Reply:
Hm OK yeah I think this may be complicating things, especially if the goal is just to avoid a test-only parameter. But I think its a valid suggestion. I'll leave this open for other reviewers, but it could also be future work. What I'd like to see is the finalize stuff merged along with this so we can release https://github.com/handshake-org/hsd/milestone/2 and users can rebuild their wallets as soon as possible. The longer we wait, the more users will have corrupted wallet state.

Reply:
I've been thinking about this and now I agree, this command could be reduced to `deep clean` and users should follow it up with a regular rescan. The only catch is the rescan MUST MUST MUST include the entire lifetime of the wallet. A normal rescan doesn't require this because it rollsback to the given height and rescans from there. `deep clean` wipes everything. 

Hopefully I get some time in the near future to clean this up and we can get another round of review to merge.

Reply:
Cool. Happy to take a crack at it, too, if it'd be helpful.